Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceCleanupRule/MSFT_IntuneDeviceCleanupRule.psm1

```diff
@@ -0,0 +1,381 @@
+        }
+
+        return $return
+    }
+    catch
+    {
+        New-M365DSCLogEntry -Message 'Error retrieving data:' `
+            -Exception $_ `
+            -Source $($MyInvocation.MyCommand.Source) `
+            -TenantId $TenantId `
+            -Credential $Credential
+
+        return $nullResult
+    }
+}
+
+function Set-TargetResource
+{
+    [CmdletBinding()]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [System.Boolean]
+        $Enabled,
+
+        [Parameter()]
+        [System.Int32]
+        [ValidateRange(30, 270)]
+        $DeviceInactivityBeforeRetirementInDays,
+
+        [Parameter()]
+        [System.String]
+        [ValidateSet('Absent', 'Present')]
+        $Ensure = 'Present',
+
+        [Parameter()]
+        [System.Management.Automation.PSCredential]
+        $Credential,
+
+        [Parameter()]
+        [System.String]
+        $ApplicationId,
+
+        [Parameter()]
+        [System.String]
+        $TenantId,
+
+        [Parameter()]
+        [System.Management.Automation.PSCredential]
+        $ApplicationSecret,
+
+        [Parameter()]
+        [System.String]
+        $CertificateThumbprint,
+
+        [Parameter()]
+        [Switch]
+        $ManagedIdentity
+    )
+
+    if ($Enabled -and $DeviceInactivityBeforeRetirementInDays -eq 0)
+    {
+        throw [System.ArgumentException]::new('DeviceInactivityBeforeRetirementInDays must be greater than 30 and less than 270 when Enabled is set to true.')
+    }
+
+    Write-Verbose -Message "Updating Device Cleanup Rule"
+
+    $ConnectionMode = New-M365DSCConnection -Workload 'MicrosoftGraph' `
+        -InboundParameters $PSBoundParameters
+
+    #Ensure the proper dependencies are installed in the current environment.
+    Confirm-M365DSCDependencies
+
+    #region Telemetry
+    $ResourceName = $MyInvocation.MyCommand.ModuleName -replace 'MSFT_', ''
+    $CommandName = $MyInvocation.MyCommand
+    $data = Format-M365DSCTelemetryParameters -ResourceName $ResourceName `
+        -CommandName $CommandName `
+        -Parameters $PSBoundParameters
+    Add-M365DSCTelemetryEvent -Data $data
+    #endregion
+
+    $url = "https://graph.microsoft.com/beta/deviceManagement/managedDeviceCleanupSettings"
+    $body = @{
+        DeviceInactivityBeforeRetirementInDays = "$(if ($Enabled) { $DeviceInactivityBeforeRetirementInDays } else { 0 })"
+    }
+    Invoke-MgGraphRequest -Method PATCH -Uri $url -Body ($body | ConvertTo-Json)
+}
+
+function Test-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [System.Boolean]
+        $Enabled,
+
+        [Parameter()]
+        [System.Int32]
+        [ValidateRange(30, 270)]
+        $DeviceInactivityBeforeRetirementInDays,
+
+        [Parameter()]
+        [System.String]
+        [ValidateSet('Absent', 'Present')]
+        $Ensure = 'Present',
```

Comment:
What is the purpose of the Ensure property in this resource. The Set method doesn't seem to be looking for it. Since there will only ever be one instance of this, shouldn't we implement the IsSingleInstance pattern instead of the Ensure one? Have a look at how we are doing it for other tenant wide settings, for example TeamsOrgWideAppsSettings

Reply:
Didn't know there exists a `IsSingleInstance` pattern. I had a look at the other resources and I hope I implemented it correctly. Please review it again, and if there is something missing, let me know.